signer/gpg2/gpg2.go

```diff
@@ -0,0 +1,247 @@
+	// e.g. 0xA2B637F535A86009
+	KeyID string
+
+	// Passphrase is the optional passphrase to use decrypt the
+	// gpg secret key
+	Passphrase string
+
+	// tmpDir is the signer's temporary working directory. It
+	// holds the gpg sec and keyrings
+	tmpDir string
+}
+
+// New initializes a pgp signer using a configuration
+func New(conf signer.Configuration) (s *GPG2Signer, err error) {
+	s = new(GPG2Signer)
+
+	if conf.Type != Type {
+		return nil, errors.Errorf("gpg2: invalid type %q, must be %q", conf.Type, Type)
+	}
+	s.Type = conf.Type
+
+	if conf.ID == "" {
+		return nil, errors.New("gpg2: missing signer ID in signer configuration")
+	}
+	s.ID = conf.ID
+
+	if conf.PrivateKey == "" {
+		return nil, errors.New("gpg2: missing private key in signer configuration")
+	}
+	s.PrivateKey = conf.PrivateKey
+
+	if conf.PublicKey == "" {
+		return nil, errors.New("gpg2: missing public key in signer configuration")
+	}
+	s.PublicKey = conf.PublicKey
+
+	if conf.KeyID == "" {
+		return nil, errors.New("gpg2: missing gpg key ID in signer configuration")
+	}
+	s.KeyID = conf.KeyID
+
+	s.Passphrase = conf.Passphrase
+
+	s.tmpDir, err = createKeyRing(s)
+	if err != nil {
+		return nil, errors.Wrap(err, "gpg2: error creating keyring")
+	}
+	return
+}
+
+// createKeyRing creates a temporary gpg sec and keyrings, loads the
+// private and public keys for the signer, and returns the temporary
+// director holding the rings
+func createKeyRing(s *GPG2Signer) (string, error) {
+	// reuse keyring in tempdir
+	prefix := fmt.Sprintf("autograph_%s_%s", s.Type, s.KeyID)
+
+	dir, err := ioutil.TempDir("", prefix)
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error creating tempdir for keyring")
+	}
+
+	// write the public key to a temp file in our signer's temp dir
+	tmpPublicKeyFile, err := ioutil.TempFile(dir, "gpg2_publickey")
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error creating tempfile for public key")
+	}
+	defer os.Remove(tmpPublicKeyFile.Name())
+	err = ioutil.WriteFile(tmpPublicKeyFile.Name(), []byte(s.PublicKey), 0755)
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error writing public key to tempfile")
+	}
+
+	// write the private key to a temp file in our signer's temp dir
+	tmpPrivateKeyFile, err := ioutil.TempFile(dir, "gpg2_privatekey")
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error creating tempfile for private key")
+	}
+	defer os.Remove(tmpPrivateKeyFile.Name())
+	err = ioutil.WriteFile(tmpPrivateKeyFile.Name(), []byte(s.PrivateKey), 0755)
+	if err != nil {
+		return "", errors.Wrap(err, "gpg2: error writing private key to tempfile")
+	}
+
+	keyRingPath := filepath.Join(dir, keyRingFilename)
+	secRingPath := filepath.Join(dir, secRingFilename)
+
+	// call gpg to create a new keyring and load the public key in it
+	gpgLoadPublicKey := exec.Command("gpg",
+		"--no-default-keyring",
+		"--keyring", keyRingPath,
+		"--secret-keyring", secRingPath,
```

Comment:
should we care about command injection via malicious config files? I'm thinking not, but the paranoid in me still thinks enforcing alphanumerical here would be good.

Reply:
Can't hurt. The paths are built from the signer Type (hardcoded) and KeyID (added alphanumeric validation in New), golang TempDir (can trust), and hardcoded file names. I suppose we could additionally validate them at each call site in case how they're constructed changes.